keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+     map<string, string> annotations = 4;
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 5;
+}
+```
+
+#### PodSandboxConfig
+
+The `PodSandboxConfig` will be supplemented with a new `class_resources` field
+that specifies the assignment of pod-level QoS-class resources. The intended
+use for this would be to be able to communicate pod-level QoS-class resource
+assignments at sandbox creation time (`RunPodSandboxRequest`).
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS resources.
++    PodQoSResources qos_resources = 10;
+ }
+
++// PodQoSResources specifies the configuration of QoS resources of a pod.
++message PodQoSResources {
++    // QoS resources the pod will be assigned to.
++    // Key-value pairs where key is name of the QoS resource and value is the
++    // name of the class.
++    map<string, string> classes = 1;
++}
+```
+
+#### ContainerStatus
+
+The `ContainerResources` message (part of `ContainerStatus`) will be extended
+to report back QoS-class resource assignments of a container, similar to other
+resources.
+
+```diff
+@@ -1251,6 +1269,8 @@ message ContainerResources {
+     LinuxContainerResources linux = 1;
+     // Resource limits configuration specific to Windows container.
+     WindowsContainerResources windows = 2;
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 3;
+```
+
+#### RuntimeStatus
+
+Extend the `RuntimeStatus` message with new `resources` field that is used to
+communicate the available QoS-class resources from the runtime to the client.
+
+This information can be used by the client (kubelet) to validate QoS-class
+resource assignments before starting a pod. In future steps kubelet will patch
+this information into node status.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
++}
+
++// ResourcesInfo contains information about the resources discovered by the
++// runtime.
++message ResourcesInfo {
++    // Pod-level QoS resources available.
++    repeated QoSResourceInfo pod_qos_resources = 1;
++    // Container-level QoS resources available.
++    repeated QoSResourceInfo container_qos_resources = 2;
++}
+
++// QoSResourceInfo contains information about one type of QoS resource.
++message QoSResourceInfo {
++    // Name of the QoS resources.
++    string Name = 1;
++    // Mutable is set to true if this resource supports in-place updates i.e.
++    // the class of a running container or sandbox can be changed.
++    bool Mutable = 2;
++    // List of classes of this QoS resource.
++    repeated QoSResourceClassInfo classes = 3;
```

Comment:
I list this under the "Open Questions" section so TBH I'm not entirely sure 🧐 I think one possibility is is to leave it as an implementation detail of the QoS resource i.e. if nothing is specified the runtime just puts it into a "default" class whetever that is. But now that you asked it this way then probably it would make sense to make it possible to inform the user about the default.

Reply:
I think there are 2 parts to this.  1) How do I (pod specifier) know what I got?  2) How do I (cluster admin) define the default?

Reply:
True.

For 1) I think I need to cook up something. Still need to think about the details. E.g. dynamic configuration updates of the QoS resources, I think it would be possible that the default is changed -> users might get wrong information if the default is changed after a pod (using default) is created.

For 2) I feel that we could rely on the QoS resource configuration on the runtime side and then use e.g. limitranges or admission webhooks to supplement that if needed.